### ISSUE TITLE
Vim 8.1 is freezing / buffering all input while another process is watching files and performing brief but I/O heavy work

### ISSUE THREAD

**Issue by nickjj** [vim/vim#3834]
Hi,

I've recently switched to Vim and I'm having difficulty getting Vim to play nicely with certain tools that watch files for changes and then perform an I/O heavy action (such as rebuilding web related files). Specifically Jekyll.

During this "rebuilding" phase which takes 3-4 seconds, all input to Vim is frozen. If I press a key, nothing happens, but the keys are being buffered because once the rebuild is done, all of the keys happen sequentially almost immediately.

I've tried disabling all plugins, and it's the same effect. I also tried using `:profile` to measure the file and func calls, and there's nothing in there causing any amount of noticeable delay.

I'm also using dedicated terminals for both Vim and the Jekyll watcher, just to eliminate any type of potential issues with `:terminal`. I'm also sure I'm not hitting `CTRL + s` (I tested for that and even disabled it).

But, here's some things I managed to figure out:

- This buffering of input for 3 seconds doesn't happen with nano
- This buffering of input for 3 seconds doesn't happen with neo vim

It seems specific to Vim.

##### My environment is:

- OS: Debian unstable (i3wm with xcompmgr)
- Terminal: rxvt-unicode (urxvt) v9.22 - released: 2016-01-23
- Vim 8.1 (terminal version)

Here's a dump from `vim --version`:

```
VIM - Vi IMproved 8.1 (2018 May 18, compiled Jan 05 2019 20:53:12)
Included patches: 1-693
Modified by team+vim@tracker.debian.org
Compiled by team+vim@tracker.debian.org
Huge version with GTK2 GUI.  Features included (+) or not (-):
+acl               +extra_search      +mouse_netterm     +tag_old_static
+arabic            +farsi             +mouse_sgr         -tag_any_white
+autocmd           +file_in_path      -mouse_sysmouse    +tcl
+autochdir         +find_in_path      +mouse_urxvt       +termguicolors
-autoservername    +float             +mouse_xterm       +terminal
+balloon_eval      +folding           +multi_byte        +terminfo
+balloon_eval_term -footer            +multi_lang        +termresponse
+browse            +fork()            -mzscheme          +textobjects
++builtin_terms    +gettext           +netbeans_intg     +textprop
+byte_offset       -hangul_input      +num64             +timers
+channel           +iconv             +packages          +title
+cindent           +insert_expand     +path_extra        +toolbar
+clientserver      +job               +perl              +user_commands
+clipboard         +jumplist          +persistent_undo   +vartabs
+cmdline_compl     +keymap            +postscript        +vertsplit
+cmdline_hist      +lambda            +printer           +virtualedit
+cmdline_info      +langmap           +profile           +visual
+comments          +libcall           -python            +visualextra
+conceal           +linebreak         +python3           +viminfo
+cryptv            +lispindent        +quickfix          +vreplace
+cscope            +listcmds          +reltime           +wildignore
+cursorbind        +localmap          +rightleft         +wildmenu
+cursorshape       +lua               +ruby              +windows
+dialog_con_gui    +menu              +scrollbind        +writebackup
+diff              +mksession         +signs             +X11
+digraphs          +modify_fname      +smartindent       -xfontset
+dnd               +mouse             +startuptime       +xim
-ebcdic            +mouseshape        +statusline        +xpm
+emacs_tags        +mouse_dec         -sun_workshop      +xsmp_interact
+eval              +mouse_gpm         +syntax            +xterm_clipboard
+ex_extra          -mouse_jsbterm     +tag_binary        -xterm_save
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -Wdate-time  -g -O2 -fdebug-prefix-map=/build/vim-SsPCMM/vim-8.1.0693=. -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -L. -Wl,-z,relro -Wl,-z,now -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E  -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -o vim   -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0 -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo -lnsl  -lselinux  -lacl -lattr -lgpm -ldl  -L/usr/lib -llua5.2 -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/x86_64-linux-gnu/perl/5.28/CORE -lperl -ldl -lm -lpthread -lcrypt  -L/usr/lib/python3.7/config-3.7m-x86_64-linux-gnu -lpython3.7m -lpthread -ldl -lutil -lm -L/usr/lib/x86_64-linux-gnu -ltcl8.6 -ldl -lz -lpthread -lm -lruby-2.5 -lpthread -lgmp -ldl -lcrypt -lm     
```

I've also talked about this issue in https://github.com/jekyll/jekyll/issues/7486, which I'm cross referencing here in case it ends up being helpful to anyone.

**Comment by mgedmin** [comment]
This is likely a result of vim trying to fsync() the swap file.  To verify, try setting `swapsync` to an empty string (check `:h 'swapsync'` for the risks inherent in that) and see if the problem goes away.

You can also try to run jekyll under `ionice -c 3` so its disk requests get lower priority than vim's and see if that works as a (safer) workaround.

**Comment by dpelle** [comment]
@mgedmin wrote:

> This is likely a result of vim trying to fsync() the swap file.

If this is indeed related to fsync(), then you may try again with this in your ~/.vimrc:
```
set nofsync
```

**Comment by nickjj** [comment]
With setting `set swapsync` in my `.vimrc` the problem still persists. I'm not sure if that's the proper syntax because now I have to hit Enter to confirm running Vim but it produces no output other than that.

If I set `set nofsync` by itself, the problem also still persists. I'm not sure if both commands need to be used together.

With neither config option in place, running `ionice -c 3` on the Jekyll watcher's PID works in the sense that it didn't produce any errors, but I'm still getting the same behavior.

So long story short, unless I'm applying `swapsync` incorrectly in the config file, none of the proposed solutions made a difference.

**Comment by nickjj** [comment]
Also, it is worth pointing out, I am using an SSD (where Jekyll is installed to) while my source code is on a HDD. If I move my source code to the SSD, I don't notice this hanging / buffering effect at all (seems fixed), and things take 1.2 seconds instead of 3, so it's still high enough where I would notice.

*[note: I would still prefer to keep the code on the HDD since it's still very speedy]*

It's also worth pointing out I've had both drives in use for years without any issues. But the thing is, it was a completely different set up where I was running Windows 10 and WSL and no Vim. Basically I'm trying to say I am pretty sure the HDD itself is ok, plus nano / neo vim with this new Debian set up confirms the drive is probably ok.

Also on the topic of drive set ups, I tried not using symlinks and the same issue persists and the HDD is an NTFS partition, which might be worth bringing up.

Here's how everything is mounted in `/etc/fstab`:

The SSD (where Jekyll is installed to):

`/dev/mapper/vg_ssd-lv_home /home           ext4    defaults,noatime        0       2`

The HDD:

`/dev/sdb2 /srv/windows/e ntfs-3g rw,uid=1000,gid=1000,dmask=0022,fmask=0133 0 0`

I also explicitly installed `inotify-tools` and the problem happened with or without that package installed.

**Comment by mgedmin** [comment]
`set swapsync` does nothing (well, it prints the current setting of 'swapsync').  You want `:set swapsync=`.

`fsync` only changes what vim does when you do an explicit `:w` or equivalent command. `swapsync` affects swap file updates that happen periodically in the background.

**Comment by nickjj** [comment]
`set swapsync=` gets rid of the error in the config, but the hanging / buffering issue remains.

**Comment by mgedmin** [comment]
> set swapsync= gets rid of the error in the config, but the hanging / buffering issue remains.

This is very interesting!  Clearly vim is blocked on I/O, but if it's not swap file updates, then where?  I wonder if strace could help.

**Comment by nickjj** [comment]
I also tried going with `set nobackup` and `set noswapfile` just to see what would happen, and it's also no difference.

strace is starting to get way above my pay grade. Would you be open to explaining exactly what I have to type to produce a meaningful output?

**Comment by nickjj** [comment]
Also, I forgot to mention this (but it was mentioned in the other issue), if I close Jekyll and try to insert and save text as fast as possible, Vim has absolutely no problem keeping up on the HDD with all plugins enabled, etc.. It seems to only be an I/O issue when Jekyll (and potentially other watchers) are set up.

**Comment by jlittlenz** [comment]
I responded to this on the vim_dev list, forgetting that some wouldn't see it there, so my apologies for the repetition.

>... an I/O heavy action ... Vim is frozen ...

This sounds much like a Linux I/O scheduler problem, especially as

>If I move my source code to the SSD, I don't notice this hanging / buffering effect

These can be inexplicably hardware dependent; identical OS and software on another machine can be problem free.  Totally frustrating because it seems not much attention was paid to it by the Linux devs for years.

For example, see https://bugs.launchpad.net/ubuntu/+source/linux/+bug/381300

Typically, deadline is used for SSDs and cfq for hard discs.

What does

    cat /sys/block/sdb/queue/scheduler

say?  If it says something like

    noop deadline [cfq]

You might try

    echo deadline | sudo tee -a  /sys/block/sdb/queue/scheduler

**Comment by nickjj** [comment]
@jlittlenz Thanks for the link. In this case it's a few small files notifying another process of changes and no USB devices are in use. Also during that 3 second Vim freeze, the rest of the system is fully responsive.

Does that still apply for me then? It was `noop deadline [cfq]` by the way.

**Comment by jlittlenz** [comment]
It applies to contention on a device with any I/O bound job that queues a lot of I/O, writes I think; other jobs using that device suffer I/O "starvation".  The "cfq" scheduler is usually the culprit, though I've seen some reports of slowness or sub-optimal throughput with "deadline".  It's quick and easy to change it on the fly with root, though one can change it permanently through several ways, like udev rules or /etc/rc.local.

**Comment by nickjj** [comment]
What would I set as the default value in case I wanted to revert the deadline addition?

**Comment by jlittlenz** [comment]
The display 

    noop deadline [cfq]

means the current value is cfq, which is presumably your system's default for a hard disc.   If you change it by writing "deadline" to the /sys pseudo-file, you can simply change it back by writing "cfq" to it, or rebooting.  Ubuntu used to have a startup script to change this for SSDs, IIRC, but I can't find it now, so perhaps the setting has moved somewhere upstream in debian or Linux.  (The system I had problems with died about 3 years ago.)

**Comment by chrisbra** [comment]
okay, closing then, as it seems to be a local configuration problem and not specific to vim.

**Comment by nickjj** [comment]
That change had no effect.

I'm not sure why this is being closed?

- Neovim works fine
- nano works fine
- emacs works fine
- VSCode works fine

Where working fine means there's no input hang up or freezing.

It's just Vim that seems to have the problem.

**Comment by chrisbra** [comment]
you might try to strace vim to see where the latency comes from. Does the build process touches the file that Vim is editing again?
Can you please come up with a reproducible example? Does it go away if you disable `fsync` setting temporarily (just to find out if fsyncing causes it). What about `:set ei=all`?

**Comment by agentzh** [comment]
I ran into a very similar problem with vim 9.1.825, which ships with Fedora 40 aarch64. Writing a small file on an SSD is very fast (the delay is unnoticeable), while doing the same on an HDD takes about 3 seconds.

The `set nofsync` command helped only a little bit, maybe less than 1 sec reduction? But the `set swapsync=` command makes the problem go away! Feels no delay now. Thanks @mgedmin for the suggestion above 5 years ago :)
